packages/@aws-cdk/aws-synthetics/lib/test-script.ts

```diff
@@ -0,0 +1,94 @@
+import * as s3 from '@aws-cdk/aws-s3';
+import { Construct } from '@aws-cdk/core';
+import { Code } from './code';
+
+/**
+ * Specify a test that the canary should run
+ */
+export class Test {
+  /**
+   * Specify a custom test with your own code
+   *
+   * @returns `Test` associated with the specified Code object
+   * @param code The script you want the canary to run
+   * @param handler The handler of the code
+   */
+  public static custom(scope: Construct, options: CustomOptions): Test {
+    const codeConfig = options.code.bind(scope);
+    return new Test({
+      handler: options.handler,
+      inlineCode: codeConfig.inlineCode,
+      s3Location: codeConfig.s3Location,
+    });
+  }
+
+  /**
+   * Use this method to access the template for heartBeat monitoring. This test will hit the specified url and report back
+   * information on the health of the website.
+   *
+   * @returns `Test` with the heartBeat template as the Code
+   * @param url The url of the website you want to test
+   *
+   * TODO: implement
+   */
+  //public static heartBeat(url: string){}
+
+  /**
+   * Use this method to access the template for api Endpoint monitoring. This test will hit the endpoint specified and report
+   * back what it receives.
+   *
+   * @param url The url of the api endpoint you want to test
+   * @param options Options to customize the api endpoint template
+   *
+   * TODO: implement
+   */
+  //public static apiEndpoint(url: string, options: apiOptions){}
+
+  /**
+   * Use this method to access the template for a broken link checker. This test will check a specified number of links and report
+   * back the first broken link found.
+   *
+   * @param url The starting url to check
+   * @param options Options to customzie the template
+   *
+   * TODO: implement
+   */
+  //public static brokenLink(url: string, options: linkOptions){}
+
+  private constructor(public readonly config: TestConfig){}
+}
+
+/**
+ * Configuration property of the test class
+ */
+export interface TestConfig {
+  /**
+   * The location of the code in S3 (mutually exclusive with `inlineCode`).
+   */
+  readonly s3Location?: s3.Location;
+
+  /**
+   * Inline code (mutually exclusive with `s3Location`).
+   */
+  readonly inlineCode?: string;
+
+  /**
+   * The handler of the code
+   */
+  readonly handler: string;
```

Comment:
Don't think you need this additional `interface` layer. Just put all these properties as public members of the `Test` class.

Also, the `Test` class is what we call a *"union"* type class. Which means its properties should represent the various static methods of the class itself. 

```suggestion
public readonly custom: CustomOptions
```

You can see an example here: https://github.com/aws/aws-cdk/blob/master/packages/@aws-cdk/custom-resources/lib/aws-custom-resource/aws-custom-resource.ts#L22

Then, in the canary resource, you do. as already mentioned:

```typescript
let codeConfig;
if (props.test.custom) {
  codeConfig = props.test.custom.code.bind(this)
} else if (props.test.brokenLink) { // when we add it...
  ...
}